SDWebImage/SDWebImagePhotosLoader.m

```diff
@@ -0,0 +1,293 @@
+}
+
+- (void)setSd_targetSize:(CGSize)sd_targetSize {
+    objc_setAssociatedObject(self, @selector(sd_targetSize), @(sd_targetSize), OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+}
+
+- (PHImageContentMode)sd_contentMode {
+    NSNumber *value = objc_getAssociatedObject(self, @selector(sd_contentMode));
+    if (value != nil) {
+        return PHImageContentModeDefault;
+    }
+    return value.integerValue;
+}
+
+- (void)setSd_contentMode:(PHImageContentMode)sd_contentMode {
+    objc_setAssociatedObject(self, @selector(sd_contentMode), @(sd_contentMode), OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+}
+
+@end
+
+@interface SDWebImagePhotosLoaderOperation : NSObject <SDWebImageOperation>
+
+@property (nonatomic, assign) PHImageRequestID requestID;
+
+@end
+
+@implementation SDWebImagePhotosLoaderOperation
+
+- (void)cancel {
+    [[PHImageManager defaultManager] cancelImageRequest:self.requestID];
+}
+
+@end
+
+@implementation SDWebImagePhotosLoader
+
++ (SDWebImagePhotosLoader *)sharedLoader {
+    static dispatch_once_t onceToken;
+    static SDWebImagePhotosLoader *loader;
+    dispatch_once(&onceToken, ^{
+        loader = [[SDWebImagePhotosLoader alloc] init];
+    });
+    return loader;
+}
+
+- (instancetype)init {
+    self = [super init];
+    if (self) {
+        PHImageRequestOptions *requestOptions = [PHImageRequestOptions new];
+        requestOptions.networkAccessAllowed = YES;
+        requestOptions.resizeMode = PHImageRequestOptionsResizeModeExact;
+        requestOptions.deliveryMode = PHImageRequestOptionsDeliveryModeFastFormat;
+        requestOptions.version = PHImageRequestOptionsVersionCurrent;
+        requestOptions.sd_targetSize = PHImageManagerMaximumSize;
+        requestOptions.sd_contentMode = PHImageContentModeDefault;
+        self.imageRequestOptions = requestOptions;
+    }
+    return self;
+}
+
+#pragma mark - Helper
++ (BOOL)isAnimatedImageWithUTType:(NSString *)UTType {
+    if (!UTType) {
+        return NO;
+    }
+    if ([UTType isEqualToString:(__bridge_transfer NSString *)kUTTypeGIF]) {
+        return YES;
+    }
+    
+    return NO;
+}
+
++ (BOOL)isPhotosStatusAuthorized {
+    return [PHPhotoLibrary authorizationStatus] == PHAuthorizationStatusAuthorized;
+}
+
+#pragma mark - SDWebImageLoader
+
+- (BOOL)canLoadWithURL:(NSURL *)url {
+    return url.sd_isPhotosURL;
+}
+
+- (id<SDWebImageOperation>)loadImageWithURL:(NSURL *)url options:(SDWebImageOptions)options progress:(SDWebImageLoaderProgressBlock)progressBlock completed:(SDWebImageLoaderCompletedBlock)completedBlock context:(SDWebImageContext *)context {
+    NSString *localIdentifier = url.sd_localIdentifier;
+    if (!localIdentifier) {
+        if (completedBlock) {
+            NSError *error = [NSError errorWithDomain:NSURLErrorDomain code:NSURLErrorFileDoesNotExist userInfo:nil];
+            completedBlock(nil, nil, error, YES);
+        }
+        return nil;
+    }
+    
+    if (![[self class] isPhotosStatusAuthorized]) {
```

Comment:
Maybe when Photos Library no authorized, we should through a exception ? Or does this authorization check impact the performance ? (Currently will check each time you request Photos asset)